File: static/js/permissions.js
/**
 * Frontend authorization module.
 *
 * Mirrors the backend policy in backend/permissions.go.
 * This is UI-only — the backend always enforces real authorization.
 * Use these helpers to show/hide UI elements and prevent unnecessary API calls.
 */

// Role constants — must match backend/models.go
export const ROLE_SYSADMIN = 'sysadmin';
export const ROLE_ADMIN = 'admin';
export const ROLE_USER = 'user';

// Action constants — must match backend/permissions.go
export const ACTION_LIST_ISSUES     = 'issue:list';
export const ACTION_GET_ISSUE       = 'issue:get';
export const ACTION_CREATE_ISSUE    = 'issue:create';
export const ACTION_UPDATE_ISSUE    = 'issue:update';
export const ACTION_DELETE_ISSUE    = 'issue:delete';
export const ACTION_ARCHIVE_ISSUE   = 'issue:archive';
export const ACTION_UNARCHIVE_ISSUE = 'issue:unarchive';
export const ACTION_MOVE_ISSUE      = 'issue:move';
export const ACTION_CREATE_TASK     = 'task:create';
export const ACTION_UPDATE_TASK     = 'task:update';
export const ACTION_DELETE_TASK     = 'task:delete';
export const ACTION_LIST_LABELS     = 'label:list';
export const ACTION_CREATE_LABEL    = 'label:create';
export const ACTION_DELETE_LABEL    = 'label:delete';
export const ACTION_LIST_USERS      = 'user:list';
export const ACTION_GET_USER        = 'user:get';
export const ACTION_CREATE_USER     = 'user:create';
export const ACTION_UPDATE_USER     = 'user:update';
export const ACTION_LIST_PROJECTS   = 'project:list';
export const ACTION_CREATE_PROJECT  = 'project:create';
export const ACTION_UPDATE_PROJECT  = 'project:update';
export const ACTION_DELETE_PROJECT  = 'project:delete';
export const ACTION_GET_STATUS_CONFIG    = 'statusconfig:get';
export const ACTION_UPDATE_STATUS_CONFIG = 'statusconfig:update';
export const ACTION_LIST_RELEASES   = 'release:list';
export const ACTION_GET_RELEASE     = 'release:get';
export const ACTION_CREATE_RELEASE  = 'release:create';
export const ACTION_UPDATE_RELEASE  = 'release:update';
export const ACTION_DELETE_RELEASE  = 'release:delete';
export const ACTION_TRIGGER_RELEASE = 'release:trigger';

// Allowlist policy — mirrors rolePermissions in backend/permissions.go.
// To add a new role or action, update both this map and the backend map.
//
// Role hierarchy:
//   ROLE_SYSADMIN: full system administration (users, projects, labels) + all ROLE_ADMIN actions
//   ROLE_ADMIN:    issue power operations (delete, archive, unarchive) + all ROLE_USER actions
//   ROLE_USER:     standard read/create/update access
const rolePermissions = {
    // Issues
    [ACTION_LIST_ISSUES]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_GET_ISSUE]:       [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_CREATE_ISSUE]:    [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_UPDATE_ISSUE]:    [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_DELETE_ISSUE]:    [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_ARCHIVE_ISSUE]:   [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_UNARCHIVE_ISSUE]: [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_MOVE_ISSUE]:      [ROLE_SYSADMIN, ROLE_ADMIN],
    // Tasks
    [ACTION_CREATE_TASK]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_UPDATE_TASK]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_DELETE_TASK]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    // Labels
    [ACTION_LIST_LABELS]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_CREATE_LABEL]:    [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_DELETE_LABEL]:    [ROLE_SYSADMIN, ROLE_ADMIN],
    // Users
    [ACTION_LIST_USERS]:      [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_GET_USER]:        [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_CREATE_USER]:     [ROLE_SYSADMIN],
    [ACTION_UPDATE_USER]:     [ROLE_SYSADMIN],
    // Projects
    [ACTION_LIST_PROJECTS]:   [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_CREATE_PROJECT]:  [ROLE_SYSADMIN],
    [ACTION_UPDATE_PROJECT]:  [ROLE_SYSADMIN],
    [ACTION_DELETE_PROJECT]:  [ROLE_SYSADMIN],
    // Status config
    [ACTION_GET_STATUS_CONFIG]:    [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_UPDATE_STATUS_CONFIG]: [ROLE_SYSADMIN, ROLE_ADMIN],
    // Releases
    [ACTION_LIST_RELEASES]:   [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_GET_RELEASE]:     [ROLE_SYSADMIN, ROLE_ADMIN, ROLE_USER],
    [ACTION_CREATE_RELEASE]:  [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_UPDATE_RELEASE]:  [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_DELETE_RELEASE]:  [ROLE_SYSADMIN, ROLE_ADMIN],
    [ACTION_TRIGGER_RELEASE]: [ROLE_SYSADMIN, ROLE_ADMIN],
};

/**
 * Returns true if the given role is permitted to perform action.
 *
 * @param {string|null} role
 * @param {string|null} action
 * @returns {boolean}
 */
export function can(role, action) {
    if (!role || !action) return false;
    const allowed = rolePermissions[action];
    if (!allowed) return false;
    return allowed.includes(role);
}

/**
 * Returns true if the given user object is permitted to perform action.
 *
 * @param {object|null} user - User object with a .role property
 * @param {string} action
 * @returns {boolean}
 */
export function userCan(user, action) {
    return can(user?.role ?? null, action);
}

File: backend/permissions.go
package backend

import "slices"

// Action represents a discrete operation subject to authorization.
// Action constants must be kept in sync with the frontend permissions.js module.
type Action string

const (
	// Issue actions
	ActionListIssues     Action = "issue:list"
	ActionGetIssue       Action = "issue:get"
	ActionCreateIssue    Action = "issue:create"
	ActionUpdateIssue    Action = "issue:update"
	ActionDeleteIssue    Action = "issue:delete"
	ActionArchiveIssue   Action = "issue:archive"
	ActionUnarchiveIssue Action = "issue:unarchive"
	ActionMoveIssue      Action = "issue:move"

	// Task actions
	ActionCreateTask Action = "task:create"
	ActionUpdateTask Action = "task:update"
	ActionDeleteTask Action = "task:delete"

	// Label actions
	ActionListLabels  Action = "label:list"
	ActionCreateLabel Action = "label:create"
	ActionDeleteLabel Action = "label:delete"

	// User management actions
	ActionListUsers  Action = "user:list"
	ActionGetUser    Action = "user:get"
	ActionCreateUser Action = "user:create"
	ActionUpdateUser Action = "user:update"

	// Project actions
	ActionListProjects  Action = "project:list"
	ActionCreateProject Action = "project:create"
	ActionUpdateProject Action = "project:update"
	ActionDeleteProject Action = "project:delete"

	// Status config actions
	ActionGetStatusConfig    Action = "statusconfig:get"
	ActionUpdateStatusConfig Action = "statusconfig:update"

	// Release actions
	ActionListReleases   Action = "release:list"
	ActionGetRelease     Action = "release:get"
	ActionCreateRelease  Action = "release:create"
	ActionUpdateRelease  Action = "release:update"
	ActionDeleteRelease  Action = "release:delete"
	ActionTriggerRelease Action = "release:trigger"
)

// rolePermissions is the single source of truth for the permission policy.
// It uses an allowlist model: a role must be explicitly listed to be granted an action.
// To add a new role, add it to the relevant actions here — no handler code changes needed.
//
// Role hierarchy:
//   - RoleSysAdmin: full system administration (users, projects, labels) + all RoleAdmin actions
//   - RoleAdmin:    issue power operations (delete, archive, unarchive) + all RoleUser actions
//   - RoleUser:     standard read/create/update access
var rolePermissions = map[Action][]UserRole{
	// Issues
	ActionListIssues:     {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionGetIssue:       {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionCreateIssue:    {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionUpdateIssue:    {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionDeleteIssue:    {RoleSysAdmin, RoleAdmin},
	ActionArchiveIssue:   {RoleSysAdmin, RoleAdmin},
	ActionUnarchiveIssue: {RoleSysAdmin, RoleAdmin},
	ActionMoveIssue:      {RoleSysAdmin, RoleAdmin},

	// Tasks
	ActionCreateTask: {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionUpdateTask: {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionDeleteTask: {RoleSysAdmin, RoleAdmin, RoleUser},

	// Labels
	ActionListLabels:  {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionCreateLabel: {RoleSysAdmin, RoleAdmin},
	ActionDeleteLabel: {RoleSysAdmin, RoleAdmin},

	// Users
	ActionListUsers:  {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionGetUser:    {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionCreateUser: {RoleSysAdmin},
	ActionUpdateUser: {RoleSysAdmin},

	// Projects
	ActionListProjects:  {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionCreateProject: {RoleSysAdmin},
	ActionUpdateProject: {RoleSysAdmin},
	ActionDeleteProject: {RoleSysAdmin},

	// Status config
	ActionGetStatusConfig:    {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionUpdateStatusConfig: {RoleSysAdmin, RoleAdmin},

	// Releases
	ActionListReleases:   {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionGetRelease:     {RoleSysAdmin, RoleAdmin, RoleUser},
	ActionCreateRelease:  {RoleSysAdmin, RoleAdmin},
	ActionUpdateRelease:  {RoleSysAdmin, RoleAdmin},
	ActionDeleteRelease:  {RoleSysAdmin, RoleAdmin},
	ActionTriggerRelease: {RoleSysAdmin, RoleAdmin},
}

// Can reports whether a user with the given role is permitted to perform action.
// Returns false for any role or action not present in the policy table (safe default).
func Can(role UserRole, action Action) bool {
	allowed, exists := rolePermissions[action]
	if !exists {
		return false
	}
	return slices.Contains(allowed, role)
}
